File: dp/cpp/timeout.cpp
// Timeout — Resilience Pattern
//
// Intent: bound the time a caller is willing to wait for an operation.
//         If the deadline passes, the caller gets an error immediately
//         instead of blocking forever.
//
// When to use:
//   - Downstream service may hang or respond slowly
//   - You must return to the user within a latency budget
//   - Pair with Retry (dp/cpp/retry.cpp) for retryable timeouts
//
// When NOT to use:
//   - The operation is local and guaranteed fast
//   - Cancellation is not possible (fire-and-forget work)

#include <iostream>
#include <future>
#include <chrono>
#include <thread>
#include <string>
#include <stdexcept>

// ============================================================================
// Generic with_timeout wrapper
// ============================================================================
template <typename F>
auto with_timeout(F&& fn, std::chrono::milliseconds deadline) -> decltype(fn()) {
    auto future = std::async(std::launch::async, std::forward<F>(fn));
    if (future.wait_for(deadline) == std::future_status::timeout) {
        throw std::runtime_error("operation timed out");
    }
    return future.get();   // may rethrow if fn threw
}

// ============================================================================
// Demo
// ============================================================================
std::string fast_service() {
    std::this_thread::sleep_for(std::chrono::milliseconds(20));
    return "fast result";
}

std::string slow_service() {
    std::this_thread::sleep_for(std::chrono::milliseconds(500));
    return "slow result";
}

int main() {
    std::cout << "=== Timeout Pattern ===\n";
    using namespace std::chrono_literals;

    // Fast call — within budget
    std::cout << "\n--- fast service (100ms budget) ---\n";
    try {
        auto r = with_timeout(fast_service, 100ms);
        std::cout << "  got: " << r << "\n";
    } catch (const std::exception& e) {
        std::cout << "  failed: " << e.what() << "\n";
    }

    // Slow call — exceeds budget
    std::cout << "\n--- slow service (100ms budget) ---\n";
    try {
        auto r = with_timeout(slow_service, 100ms);
        std::cout << "  got: " << r << "\n";
    } catch (const std::exception& e) {
        std::cout << "  failed: " << e.what() << "\n";
    }

    // Give the detached async thread time to finish to avoid sanitizer warnings
    std::this_thread::sleep_for(500ms);

    std::cout << "\nKey points:\n";
    std::cout << " * Prevents indefinite blocking on slow/hung operations\n";
    std::cout << " * Caller regains control immediately on timeout\n";
    std::cout << " * The timed-out task may still complete in the background\n";
    std::cout << " * For cooperative cancellation, combine with stop_token (C++20)\n";
    return 0;
}

File: dp/rust/src/timeout.rs
// Timeout Pattern in Rust
// Bound the execution time of operations

use std::sync::mpsc;
use std::thread;
use std::time::Duration;

fn with_timeout<F, T>(duration: Duration, f: F) -> Result<T, String>
where
    F: FnOnce() -> T + Send + 'static,
    T: Send + 'static,
{
    let (tx, rx) = mpsc::channel();

    thread::spawn(move || {
        let result = f();
        let _ = tx.send(result);
    });

    rx.recv_timeout(duration)
        .map_err(|_| format!("Operation timed out after {:?}", duration))
}

fn slow_operation(millis: u64) -> String {
    thread::sleep(Duration::from_millis(millis));
    format!("Completed after {}ms", millis)
}

fn main() {
    // Fast operation — succeeds within timeout
    println!("=== Fast Operation ===");
    match with_timeout(Duration::from_millis(200), || slow_operation(50)) {
        Ok(result) => println!("Success: {}", result),
        Err(e) => println!("Failed: {}", e),
    }

    // Slow operation — exceeds timeout
    println!("\n=== Slow Operation ===");
    match with_timeout(Duration::from_millis(100), || slow_operation(500)) {
        Ok(result) => println!("Success: {}", result),
        Err(e) => println!("Failed: {}", e),
    }

    // Multiple operations with different timeouts
    println!("\n=== Multiple Timeouts ===");
    let operations = vec![
        ("Quick", 30u64, 100u64),
        ("Medium", 80, 100),
        ("Slow", 200, 100),
    ];

    for (name, work_ms, timeout_ms) in operations {
        let timeout = Duration::from_millis(timeout_ms);
        let result = with_timeout(timeout, move || slow_operation(work_ms));
        match result {
            Ok(r) => println!("[{}] {}", name, r),
            Err(e) => println!("[{}] {}", name, e),
        }
    }

    thread::sleep(Duration::from_millis(300)); // Let leaked threads finish
}
